docs/rfcs/rfc-035-read-write-controls.md

```diff
@@ -0,0 +1,378 @@
+  name: String!
+  comments: [Comment!]! @relationship(type: "HAS_AUTHOR", direction: IN) @aggregations
+  posts: [Post!]! @relationship(type: "HAS_AUTHOR", direction: IN) @aggregations
+}
+
+type Post @write(operations: [CREATE, UPDATE, DELETE]) {
+  content: String!
+  comments: [Comment!]! @relationship(type: "HAS_COMMENT", direction: OUT) @write(operations: [CREATE])
+  author: User! @relationship(type: "HAS_AUTHOR", direction: OUT) @write(operations: [CONNECT])
+}
+
+type Comment @write(operations: [UPDATE, DELETE]) {
+  content: String!
+  post: Post! @relationship(type: "HAS_COMMENT", direction: IN)
+  author: User! @relationship(type: "HAS_AUTHOR", direction: OUT) @write(operations: [CONNECT])
+}
+
+extend schema @readonly
+```
+
+* `extend schema @readonly` makes the entire schema read-only
+* The following Mutation Fields are added:
+  * `createUsers`, `updateUsers`
+  * `createPosts`, `updatePosts`, `deletePosts`
+  * `updateComments`, `deleteComments`
+* All relationships _from_ `User` are read-only, but with added aggregation fields
+* `Comment` nodes can _only_ be created when coming from a `Post`
+* The `post` field from `Comment` is read-only
+* The `author` field _must_ be connected to a `User` when coming from either a `Post` or `Comment`
+
+This highlights some disadvantages:
+
+* Despite `CREATE`, `UPDATE` and `CONNECT` being specified on `User`, we ended up overwriting them for all relationships pointing to `User`
+
+## Solution
+
+It is likely that a single directive for this functionality is no longer desired, given the ever-expanding capabilities of the library.
+
+This RFC will discuss some options on how functionality might be split.
+
+### Globals
+
+There are currently `@readonly` and `@writeonly` directive which are only applicable to fields. These will be extended:
+
+```gql
+directive @readonly on SCHEMA | OBJECT | FIELD_DEFINITION
+directive @writeonly on SCHEMA | OBJECT | FIELD_DEFINITION
+```
+
+Only one of the above can be used in any given location.
+
+To disable all Mutations (and thus, all nested operations):
+
+```gql
+extend schema @readonly
+```
+
+To disable all Queries, except for the Relay `node` Query:
+
+```gql
+extend schema @writeonly
+```
+
+To disable all nested write operations across a particular relationship:
+
+```gql
+type Movie {
+  title: String!
+}
+
+type Actor {
+  name: String!
+  movies: [Movie!]! @relationship(type: "ACTED_IN", direction: OUT) @readonly
+}
+```
+
+*OPEN QUESTION*: given the following, how would you re-enable reads for the `Actor.movies` relationship?
+
+```gql
+type Movie @writeonly {
+  title: String!
+}
+
+type Actor {
+  name: String!
+  movies: [Movie!]! @relationship(type: "ACTED_IN", direction: OUT)
+}
+```
+
+One proposal would be to introduce a directive mirroring `@write`, which could be used to re-able read permissions in a particular location:
+
+```gql
+directive @read (
+  enabled: Boolean! = true
+) on OBJECT | FIELD_DEFINITION
```

Comment:
`@read(enabled: false)` on `SCHEMA` is essentially `@writeonly` - we should think about the existence of all 4 of these directives.

Reply:
I stand by all 4 directives existing - at the very least, `@readonly` and `@writeonly` are applicable to _all_ field definitions, whilst `@write` is only for use on relationship fields.

The overhead is not in how many directives we have, but how we parse all of them. Once we sort that out, I don't think we'll worried about this - at least for directives which make sense which I think all of these do.